google-cloud-storage/src/main/java/com/google/cloud/storage/SignatureInfo.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Signature Info holds payload components of the string that requires signing.
+ * 
+ * @see <a href=
+ *      "https://cloud.google.com/storage/docs/access-control/signed-urls#string-components">
+ *      Components</a>
+ */
+public class SignatureInfo {
+
+  private static final char COMPONENT_SEPARATOR = '\n';
+  private static final char HEADER_SEPARATOR = ':';
+
+  private final HttpMethod httpVerb;
+  private final String contentMd5;
+  private final String contentType;
+  private final long expiration;
+  private final Map<String, String> canonicalizedExtensionHeaders;
+  private final URI canonicalizedResource;
+
+  private SignatureInfo(Builder builder) {
+    this.httpVerb = builder.httpVerb;
+    this.contentMd5 = builder.contentMd5;
+    this.contentType = builder.contentType;
+    this.expiration = builder.expiration;
+    this.canonicalizedExtensionHeaders = builder.canonicalizedExtensionHeaders;
+    this.canonicalizedResource = builder.canonicalizedResource;
+  }
+
+  /**
+   * Constructs payload to be signed.
+   * 
+   * @return paylod to sign
+   * @see <a href="https://cloud.google.com/storage/docs/access-control#Signed-URLs">Signed URLs</a>
+   */
+  public String constructUnsignedPayload() {
+    StringBuilder payload = new StringBuilder();
+
+    payload.append(httpVerb.name()).append(COMPONENT_SEPARATOR);
+    if (contentMd5 != null) {
+      payload.append(contentMd5);
+    }
+    payload.append(COMPONENT_SEPARATOR);
+    
+    if (contentType != null) {
+      payload.append(contentType);
+    }
+    payload.append(COMPONENT_SEPARATOR);
+    
+    payload.append(expiration).append(COMPONENT_SEPARATOR);
+    
+    if (canonicalizedExtensionHeaders != null) {
+      
+      List<String> orderedKeys = new ArrayList<>(canonicalizedExtensionHeaders.keySet());
+      Collections.sort(orderedKeys);
```

Comment:
Are the headers just ordered alphabetically?

Reply:
Work in progress.

[Implementing](https://cloud.google.com/storage/docs/access-control/signed-urls#about-canonical-extension-headers).

Reply:
Done